354 add prometheus metrics

Ayreon2084 opened the pull request:
Add Prometheus metrics:

1) "impulse_messenger_api_request_duration_seconds" Histogram metric for RateLimitedClient.request().

Implemented as an async context manager to:
- measure HTTP request duration (buckets: 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0, +Inf);
- record messenger_type, status (1-5xx, no_response) and error (none, connection_failed, timeout, unknown) labels .

2) "impulse_status" Gauge metric.

Check whether instance is primary (1.0) or standby (0.0). Include /metrics endpoint to StandbyMiddleware.dispatch() to let this endpoint be available when in standby mode.

3) "impulse_queue_latency_seconds" Gauge metric for AsyncQueue.get_next_ready_item().

Calculate delay between current time and QueueItem.datetime when item is retrieved from queue.

Ayreon2084 committed:
Add GET /metrics endpoint, update requirements.txt and gitignore

Ayreon2084 committed:
Add api_response_time_seconds metrics

Ayreon2084 committed:
Add impulse_status gauge metric, let /metrics endpoint be available when in standby mode

Ayreon2084 committed:
Add impulse_queue_delay_seconds gauge metric

Ayreon2084 committed:
Rename metrics, instead of 'status_class' label made 'status' and 'error' labels in 'impulse_messenger_api_request_duration_seconds' metric

tansdf commented on app/http_client/rate_limited_client.py:
```diff
@@ -78,10 +80,12 @@ def __init__(
     ):
         self.rate_limit = rate_limit
         self.rate_window = rate_window
+        self.messenger = messenger or 'none'
```
What's the point creating `Optional` parameter with default `None` and always setting it to `'none'`?

Ayreon2084 replied:
Sure. Here and after RateLimitedClient won't have any attributes related to another classes.

Completely forgot about get_config() providing access to messenger_type from impulse.yml.

tansdf commented on app/http_client/rate_limited_client.py:
```diff
@@ -78,10 +80,12 @@ def __init__(
         connector_limit: int = 100,
-        connector_limit_per_host: int = 30
+        connector_limit_per_host: int = 30,
+        messenger: Optional[str] = None
```
You're putting a `messenger` parameter here. In our code messenger is the Application class, not a simple string, so the naming is incorrect.
Also you're putting the `messenger` into the https client class. What for?

Ayreon2084 replied:
Ok.

tansdf commented on app/http_client/rate_limited_client.py:
```diff
@@ -209,7 +213,31 @@ async def request(self, method: str, url: str, **kwargs):
+
+            except Exception as e:
+                logger.error(f"Request to {url} failed: {type(e).__name__}: {e}")
+                raise
```
That's a horrible implementation. You've created a context manager function to just wrap the request method. It doesn't make any sense to catch errors and assign them to the context rerising it.
Why don't you make a simple decorator?

Ayreon2084 replied:
Implemented decorator instead.

tansdf commented on app/im/application.py:
```diff
@@ -312,7 +312,8 @@ def _setup_http(self) -> RateLimitedClient:
             connector_limit=100,
-            connector_limit_per_host=30
+            connector_limit_per_host=30,
+            messenger=self.type.value  # Messenger type for metrics
```
What for we're putting the messenger type to metrics?

Ayreon2084 replied:
So far there's just one opportunity: if multiple applications are running on different servers with different messengers, this label allows us to clearly distinguish which instance the metric belongs to.

tansdf commented on app/metrics.py:
```diff
@@ -0,0 +1,46 @@
+            messenger=messenger,
+            status=status,
+            error=error
+        ).observe(duration)
```
No that's not the way. We can't allow the existance of context manager function that collects performance metrics. It must be a singleton MetricsCollector class that implements the needed decorator

Ayreon2084 replied:
Got it.

tansdf commented on app/http_client/rate_limited_client.py:
```diff
@@ -209,7 +213,31 @@ async def request(self, method: str, url: str, **kwargs):
+            except asyncio.TimeoutError as e:
+                context['error'] = 'timeout'
+                logger.error(f"Request to {url} timed out: {type(e).__name__}: {e}")
+                raise e
```
Rerising `e` as `raise e` brakes the original traceback

Ayreon2084 replied:
Missed it. My fault.

tansdf commented on app/queue/queue.py:
```diff
@@ -80,6 +81,8 @@ async def get_next_ready_item(self) -> Optional[Tuple[str, str, str, Any]]:
             if self._items and self._items[0].datetime <= now:
                 item = self._items.pop(0)
+                delay = (now - item.datetime).total_seconds()
+                metrics_queue_latency_seconds.set(delay)
```
The `metrics_queue_latency_seconds` is only set. Never reset

Ayreon2084 replied:
Your comment gave us some food for thought, and we decided to refactor the behavior of this metric. Instead of calculating the delay when an item is retrieved from the queue, we now calculate it on each request to the /metrics endpoint.
This provides more predictable data: a zero value now reliably means either that the queue is empty or that all scheduled datetimes are in the future relative to the time of the metrics request.

tansdf commented on app/http_client/rate_limited_client.py:
```diff
@@ -209,7 +213,31 @@ async def request(self, method: str, url: str, **kwargs):
+                status_code = response.status
+                context['status'] = f'{status_code // 100}xx'
+                context['error'] = 'none'
+                logger.info(f"Request to {url} completed with status {status_code} ({context['status']})")
```
Too much uselsess logs

Ayreon2084 replied:
I still can’t feel when logging is sufficient and when it’s excessive.

tansdf commented on main.py:
```diff
@@ -360,6 +365,19 @@ async def websocket_endpoint(websocket: WebSocket):
+    return Response(
+        content=generate_latest(),
+        media_type=CONTENT_TYPE_LATEST
+    )
```
It should be produced by the MetricsCollector class. Endpoint should not encapsulate such logic

Ayreon2084 replied:
Done.

Ayreon2084 committed:
refactor(metrics): refactor metrics and move queue_latency_seconds update to /metrics endpoint

Ayreon2084 committed:
fix[metrics] Remove redundant except Exception in measure_request decorator

tansdf commented on app/http_client/rate_limited_client.py:
```diff
@@ -209,32 +212,53 @@ async def request(self, method: str, url: str, **kwargs):
+            logger.error(
+                f"Request to {url} failed: {type(e).__name__}: {e}"
+            )
+            raise
```
Why do we need these additional logs?

Ayreon2084 replied:
Removed these logs.

tansdf commented on app/metrics.py:
```diff
@@ -0,0 +1,150 @@
+    metrics_collector.messenger_api_request_duration_seconds
+)
+metrics_status = metrics_collector.status
+metrics_queue_latency_seconds = metrics_collector.queue_latency_seconds
```
Why do we proxy access to the MetricsCollector properties over the additional variables?

tansdf commented on app/metrics.py:
```diff
@@ -0,0 +1,150 @@
+
+    Decorator function that wraps MetricsCollector.measure_request.
+    """
+    return metrics_collector.measure_request(func)
```
Why do we have this additional function istead of the direct accesss to the MetricsCollector.measure_request? It xould be a classmethod that checks the instance existance creates new if needed and then mesures what it needs to measure

Ayreon2084 replied:
Good point. I've refactored the code accordingly:

- Removed the module-level "MetricsCollector" instance and the wrapper function;
- Converted "measure_request" to a @classmethod that can be used as a decorator;
- The "MetricsCollector" instance is now initialized during lifespan startup and stored in "app.state.metrics"

tansdf commented on app/metrics.py:
```diff
@@ -0,0 +1,150 @@
+
+        return wrapper
+
+    async def update_queue_latency(self, queue):
```
I don't like the place you collect these metrics, why don't add the collecting to the QueueManager, where we can just compare the item datetime with current and put it into the metric. This way we will just get the metric at trigger moment, but shouldn't we collect statistics, like avg, min, max about the whole processing?

Ayreon2084 replied:
@DiTsi decided that current implementation is more that adequate.

tansdf commented on app/metrics.py:
```diff
@@ -0,0 +1,150 @@
+metrics_queue_latency_seconds = metrics_collector.queue_latency_seconds
+
+
+def measure_request(func):
```
Is it used anywhere?

tansdf commented on main.py:
```diff
@@ -360,6 +364,16 @@ async def websocket_endpoint(websocket: WebSocket):
+@metrics_router.get("/metrics")
+async def get_metrics(request: Request):
+    """Prometheus metrics endpoint"""
+    queue = getattr(request.app.state, 'queue', None)
```
Using getattr is not a good idea here. The queue should always exist and if not we need to handle it here.

Ayreon2084 replied:
To avoid None checks in the metrics layer and guarantee that both the queue and the metrics collector instances always exist, I initialized both of them in the FastAPI lifespan.

If the queue is empty or there are no items ready to be processed at the moment of the /metrics request, queue_latency_seconds is set to 0. Otherwise, the delay is calculated based on the timestamp of the first queued item.

Ayreon2084 committed:
refactor[metrics]: use singleton via app.state.metrics

- Remove module-level MetricsCollector instance;
- Initialize app.state.metrics with during FastAPI lifespan startup;
- Convert measure_request decorator to classmethod;
- Initialize app.state.queue with an empty AsyncQueue() during FastAPI lifespan startup (standby mode no longer results in queue=None);
- Remove None checks from metrics as long as queue is now guaranteed to exist;
- Simplify "/metrics" endpoint: always pass a valid queue to generate_metrics_response;
- Remove additional logs from RateLimitedClient.request.

Ayreon2084 committed:
refactor(metrics): remove redundant try/except since errors are handled in the decorator

Ayreon2084 committed:
refactor: remove redundant response var

DiTsi committed:
Refactoring: (#381)

- added HTTP_PREFIX
- removed standard metrics
- removed MetricsCollector class
